src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -342,156 +332,173 @@ async Task<bool> NextResult(bool async, bool isConsuming=false)
-                        pStatement?.CompleteUnprepare();
-                        throw;
-                    }
+                        Expect<BindCompleteMessage>(await Connector.ReadMessage(async), Connector);
+                        msg = await Connector.ReadMessage(async);
 
-                    Expect<BindCompleteMessage>(await Connector.ReadMessage(async), Connector);
-                    msg = await Connector.ReadMessage(async);
+                        RowDescription = statement.Description = msg.Code switch
+                        {
+                            BackendMessageCode.NoData => null,
 
-                    RowDescription = statement.Description = msg.Code switch
-                    {
-                        BackendMessageCode.NoData => null,
+                            // RowDescription messages are cached on the connector, but if we're auto-preparing, we need to
+                            // clone our own copy which will last beyond the lifetime of this invocation.
+                            BackendMessageCode.RowDescription => pStatement == null
+                                ? (RowDescriptionMessage)msg
+                                : ((RowDescriptionMessage)msg).Clone(),
 
-                        // RowDescription messages are cached on the connector, but if we're auto-preparing, we need to
-                        // clone our own copy which will last beyond the lifetime of this invocation.
-                        BackendMessageCode.RowDescription => pStatement == null
-                            ? (RowDescriptionMessage)msg
-                            : ((RowDescriptionMessage)msg).Clone(),
+                            _ => throw Connector.UnexpectedMessageReceived(msg.Code)
+                        };
 
-                        _ => throw Connector.UnexpectedMessageReceived(msg.Code)
-                    };
+                        pStatement?.CompletePrepare();
+                    }
 
-                    pStatement?.CompletePrepare();
-                }
+                    if (RowDescription == null)
+                    {
+                        // Statement did not generate a resultset (e.g. INSERT)
+                        // Read and process its completion message and move on to the next statement
 
-                if (RowDescription == null)
-                {
-                    // Statement did not generate a resultset (e.g. INSERT)
-                    // Read and process its completion message and move on to the next statement
+                        msg = await ReadMessage(async);
+                        switch (msg.Code)
+                        {
+                        case BackendMessageCode.CompletedResponse:
+                        case BackendMessageCode.EmptyQueryResponse:
+                            break;
+                        default:
+                            throw Connector.UnexpectedMessageReceived(msg.Code);
+                        }
```

Comment:
```csharp
if (msg.Code != BackendMessageCode.CompletedResponse &&
    msg.Code != BackendMessageCode.EmptyQueryResponse)
{
    throw Connector.UnexpectedMessageReceived(msg.Code);
}
```

Reply:
As above, I use switches for these cases mainly because of (theoretical) perf.